src/Agent.Listener/Configuration/ConfigurationProvider.cs

```diff
@@ -0,0 +1,357 @@
+
+        public void InitConnection(IAgentServer agentServer)
+        {
+            InitializeServerConnection(agentServer);
+        }
+
+        public void UpdateAgentSetting(AgentSettings settings)
+        {
+            // No implementation required
+        }
+
+        public string GetServerUrl(CommandSettings command)
+        {
+            return command.GetUrl(false);
+        }
+
+        public async Task<int> GetPoolId(CommandSettings command)
+        {
+            int poolId = 0;
+            string poolName;
+            while (true)
+            {
+                poolName = command.GetPool();
+                try
+                {
+                    poolId = await GetPoolIdAsync(poolName);
+                }
+                catch (Exception e) when (!command.Unattended)
+                {
+                    _term.WriteError(e);
+                }
+
+                if (poolId > 0)
+                {
+                    break;
+                }
+
+                _term.WriteError(StringUtil.Loc("FailedToFindPool"));
+            }
+            return poolId;            
+        }
+
+        public Task<TaskAgent> UpdateAgentAsync(int poolId, TaskAgent agent)
+        {
+            return UpdateAgent(poolId, agent);
+        }
+
+        public Task<TaskAgent> AddAgentAsync(int poolId, TaskAgent agent)
+        {
+            return AddAgent(poolId, agent);
+        }
+
+        public Task DeleteAgentAsync(int agentPoolId, int agentId)
+        {
+            return DeleteAgent(agentPoolId,agentId);
+        }
+
+        public async Task<IAgentServer> TestConnectAsync(string url, VssCredentials creds)
+        {
+            await TestConnectionAsync(url, creds);
+            return _agentServer;
+        }
+
+        private async Task<int> GetPoolIdAsync(string poolName)
+        {
+            int poolId = 0;
+            List<TaskAgentPool> pools = await _agentServer.GetAgentPoolsAsync(poolName);
+            Trace.Verbose("Returned {0} pools", pools.Count);
+
+            if (pools.Count == 1)
+            {
+                poolId = pools[0].Id;
+                Trace.Info("Found pool {0} with id {1}", poolName, poolId);
+            }
+
+            return poolId;
+        }
+
+    }
+
+    public sealed class MachineGroupAgentConfigProvider : ConfigurationProvider, IConfigurationProvider
+    {
+        private IAgentServer _collectionAgentServer =null;
+        private string _projectName;
+        private string _collectionName;
+        private string _machineGroupName;
+        private string _serverUrl;
+        private bool _isHosted = false;
+
+        public string ConfigurationProviderType
+            => Constants.Agent.AgentConfigurationProvider.DeploymentAgentConfiguration;
+
+        public void InitConnection(IAgentServer agentServer)
+        {
+            InitializeServerConnection(agentServer);
+        }
+
+        public string GetServerUrl(CommandSettings command)
```

Comment:
In case of on-prem scenario it is 3 inputs - collection, project and machineGroup name. In case of VSTS scenario it is 2 project and machineGroupName. 
The most of logic is only of on-prem scenario, which may go away in future sprints, for VSTS case the logic is very much simple